libmisc/prefix_flag.c

```diff
@@ -0,0 +1,342 @@
+ *
+ * This shall be called before accessing the passwd, group, shadow,
+ * gshadow, useradd's default, login.defs files (non exhaustive list)
+ * or authenticating the caller.
+ *
+ * The audit, syslog, or locale files shall be open before
+ */
+extern const char* process_prefix_flag (const char* short_opt, int argc, char **argv)
+{
+	/*
+	 * Parse the command line options.
+	 */
+	int i;
+	const char *prefix = NULL;
+
+	for (i = 0; i < argc; i++) {
+		if (   (strcmp (argv[i], "--prefix") == 0)
+		    || (strcmp (argv[i], short_opt) == 0)) {
+			if (NULL != prefix) {
+				fprintf (stderr,
+				         _("%s: multiple --prefix options\n"),
+				         Prog);
+				exit (E_BAD_ARG);
+			}
+
+			if (i + 1 == argc) {
+				fprintf (stderr,
+				         _("%s: option '%s' requires an argument\n"),
+				         Prog, argv[i]);
+				exit (E_BAD_ARG);
+			}
+			prefix = argv[i + 1];
+		}
+	}
+
+	
+
+	if (prefix != NULL) {
+		if ( prefix[0] == '\0' || !strcmp(prefix, "/"))
+			return ""; /* if prefix is "/" then we ignore the flag option */
+		/* should we prevent symbolic link from being used as a prefix? */
+
+		size_t len;
+		len = strlen(prefix) + strlen(PASSWD_FILE) + 2;
+		passwd_db_file = xmalloc(len);
+		snprintf(passwd_db_file, len, "%s/%s", prefix, PASSWD_FILE);
+		pw_setdbname(passwd_db_file);
+
+		len = strlen(prefix) + strlen(GROUP_FILE) + 2;
+		group_db_file = xmalloc(len);
+		snprintf(group_db_file, len, "%s/%s", prefix, GROUP_FILE);
+		gr_setdbname(group_db_file);
+	
+#ifdef  SHADOWGRP
+		len = strlen(prefix) + strlen(SGROUP_FILE) + 2;
+		sgroup_db_file = xmalloc(len);
+		snprintf(sgroup_db_file, len, "%s/%s", prefix, SGROUP_FILE);
+		sgr_setdbname(sgroup_db_file);
+#endif
+#ifdef	USE_NIS
+		__setspNIS(0); /* disable NIS for now, at least until it is properly supporting a "prefix" */
+#endif
+
+		len = strlen(prefix) + strlen(SHADOW_FILE) + 2;
+		spw_db_file = xmalloc(len);
+		snprintf(spw_db_file, len, "%s/%s", prefix, SHADOW_FILE);
+		spw_setdbname(spw_db_file);
+		
+
+		len = strlen(prefix) + strlen("/etc/subuid") + 2;
+		suid_db_file = xmalloc(len);
+		snprintf(suid_db_file, len, "%s/%s", prefix, "/etc/subuid");
+		sub_uid_setdbname(suid_db_file);
+
+		len = strlen(prefix) + strlen("/etc/subgid") + 2;
+		sgid_db_file = xmalloc(len);
+		snprintf(sgid_db_file, len, "%s/%s", prefix, "/etc/subgid");
+		sub_gid_setdbname(sgid_db_file);
+
+		len = strlen(prefix) + strlen("/etc/login.defs") + 2;
+		def_conf_file = xmalloc(len);
+		snprintf(def_conf_file, len, "%s/%s", prefix, "/etc/login.defs");
+		setdef_config_file(def_conf_file);
+	}	
+
+	if (prefix == NULL)
+		return "";
+	return prefix;
+}
+
+
+extern struct group *prefix_getgrnam(const char *name)
+{
+	if (group_db_file) {
+		FILE* fg;
+		struct group * grp = NULL;
+
+		fg = fopen(group_db_file, "rt");
+		if(!fg)
+			return NULL;
+		while(grp = fgetgrent(fg)) {
+			if(!strcmp(name, grp->gr_name))
+				break;
+		}
+		fclose(fg);
+		return grp;
+	}
```

Comment:
the } else is not needed here, since the if block always return.  (same in the ones below).

Reply:
ok. I will add some checks.